pkg/convert/pprof/streaming/parser_streaming.go

```diff
@@ -0,0 +1,521 @@
+package streaming
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"github.com/pyroscope-io/pyroscope/pkg/storage"
+	"github.com/pyroscope-io/pyroscope/pkg/storage/metadata"
+	"github.com/pyroscope-io/pyroscope/pkg/storage/segment"
+	"github.com/pyroscope-io/pyroscope/pkg/storage/tree"
+	"github.com/richardartoul/molecule"
+	"github.com/richardartoul/molecule/src/codec"
+	"github.com/valyala/bytebufferpool"
+	"io"
+	"time"
+)
+
+type ParserConfig struct {
+	Putter        storage.Putter
+	SpyName       string
+	Labels        map[string]string
+	SkipExemplars bool
+	SampleTypes   map[string]*tree.SampleTypeConfig
+	//StackFrameFormatter StackFrameFormatter
+}
+
+type MoleculeParser struct {
+	putter        storage.Putter
+	spyName       string
+	labels        map[string]string
+	skipExemplars bool
+	sampleTypes   map[string]*tree.SampleTypeConfig
+	//stackFrameFormatter pprof.StackFrameFormatter
+
+	sampleTypesFilter func(string) bool
+
+	previousCache LabelsCache
+
+	startTime time.Time
+	endTime   time.Time
+	ctx       context.Context
+
+	profile             []byte
+	strings             [][]byte
+	profileIDLabelIndex int
+	sampleTypesParsed   []valueType
+	periodType          valueType
+	period              int
+	mainBuf             *codec.Buffer
+	tmpBuf1             *codec.Buffer
+	tmpBuf2             *codec.Buffer
+
+	functions []function
+	locations []location
+
+	indexes []int
+	types   []int
+
+	tmpValues []int64
+	tmpLabels []label
+	tmpStack  [][]byte
+
+	finder Finder
+}
+
+func NewStreamingParser(config ParserConfig) *MoleculeParser {
+	//if config.StackFrameFormatter == nil {//todo
+	//	config.StackFrameFormatter = &pprof.UnsafeFunctionNameFormatter{}
+	//}
+	return &MoleculeParser{
+		putter:        config.Putter,
+		spyName:       config.SpyName,
+		labels:        config.Labels,
+		sampleTypes:   config.SampleTypes,
+		skipExemplars: config.SkipExemplars,
+
+		previousCache:     make(LabelsCache),
+		sampleTypesFilter: filterKnownSamples(config.SampleTypes),
+	}
+}
+
+func (p *MoleculeParser) GetSampleTypesFilter() func(string) bool {
+	return p.sampleTypesFilter
+}
+
+func (p *MoleculeParser) SetSampleTypesFilter(f func(string) bool) {
+	p.sampleTypesFilter = f
+}
+func (p *MoleculeParser) ParsePprof(ctx context.Context, startTime, endTime time.Time, bs []byte) error {
+	p.startTime = startTime
+	p.endTime = endTime
+	p.ctx = ctx
+	defer func() { p.ctx = nil }()
+	if len(bs) < 2 {
+		return fmt.Errorf("failed to read pprof profile header")
+	}
+	if bs[0] == 0x1f && bs[1] == 0x8b {
+		gzipr, err := gzip.NewReader(bytes.NewReader(bs))
+		if err != nil {
+			return fmt.Errorf("failed to create pprof profile zip reader: %w", err)
+		}
+		defer gzipr.Close()
+		buf := bytebufferpool.Get()
+		defer bytebufferpool.Put(buf)
+		if _, err = io.Copy(buf, gzipr); err != nil {
+			return err
+		}
+		p.profile = buf.Bytes()
+		defer func() { p.profile = nil }()
```

Comment:
the field is to avoid passing the buffer around multiple functions , there is no need to keep profile after parsing is done since it may contain garbage. I remove a reference to it just to prevent possible access to the garbage in the future. I also rewrite without defer. OK?